### PR TITLE
Add Backend

### PR DESCRIPTION
**Pull request by StephenODea54**
Added the backend for app. It is built using:
  - Node + Express for the Rest API
  - TypeScript
  - Prisma for the ORM
  - PlanetScale for the DB (It uses MySQL under the hood which is 🗑 but the free tier for Planet Scale is booonkers).

For the schema we have three tables (technically four, but the one table (`_TagsAndPosts`) is just a join table. Below is a janky ER diagram illustrating the tables with their columns, keys, names, and types. It's janky because I did not include the relationship indicators (too lazy), so I will write them about instead:
  - A user has many posts
  - A post belongs to one and only one user
  - A post can contain many tags
  - A tag can be contained in many posts

image

For the actual API, there are the routes:
  - `/`
    - Entry Point, displays a welcome message
  - `/posts`
    - GET: Displays all blog posts
    - POST: Creates a new blog post (must provide title, content, authorName, and tags as arguments)
  - `/post/:id`
    - `:id` is the query param
    - PUT: Updates an existing blog post by the id (must provide title, content, and tags as arguments)
    - DELETE: Deletes an existing blog post by the id
  - `/tags`
    - GET: Displays all tags

I did not add any routes for the Users because I'm still unsure as to how we are going to go about authentication. Regardless, I don't anticipate this to be a hard task.

Also also, I did not integrate any of the API stuff in the front end because it looks like we got a lot of branches and I want to get this stuff in before I make additional changes on the UI using this API.

### CODE REVIEW

**Review by jpatterson933 — APPROVED** [review]
I'll need to read up on the documentation for the specifics, but glad we have ourselves a first draft!!

**Comment by StephenODea54** [comment]
@jpatterson933 sounds good! Feel free to hmu with any questions about it! It's obviously subject to change as we integrate more things on the front end. It could also use some cleaning up (really just wanted to get first pass through). Once this gets merged I'll create some tickets for other folks to contribute on it.

**Comment by nathanjslim** [comment]
This is great, especially for the amount of time you took man! 

Looks like there's a few syntax errors, at least in Blog.jsx that's preventing functionality. 
Think merging to main at this stage will break the current blog we have. 

Also not sure if this is caused by the pull/checkout you probably did to catch your branch up to main but Blog.jsx has two arrow functions called 'Blog'.. should I rewrite the search function off yours?

File: client/src/pages/Blog.jsx [hunk 1]
```diff
@@ -1,3 +1,16 @@
...
+  return(
+    <section style={{ whiteSpace: 'pre-line' }}>
+      <h1 id={`blog-id-${id}`}>{title}</h1>
+      <h3>{createdAt} | <span>{author}</span></h3>
+      <p>{content}</p>
+    </section>
+  );
+};
+
+const Blog = () => {
```
Comment (nathanjslim) [inline client/src/pages/Blog.jsx]: Duplicate 'Blog' function

File: client/src/pages/Blog.jsx [hunk 2]
```diff
@@ -15,109 +28,19 @@ const Blog = ({ visibleElements }) => {
...
-            <br></br>
-            With the Eye of the Storm in their possession, the adventurers were able to harness the power of the elements themselves. They used the artifact to call forth a great storm, the likes of which the dragons had never seen.
-            As the storm raged on, the dragons were weakened and finally defeated, their power absorbed by the adventurers. Bruinbrewinbeer knew that he had found his true calling, as a brewer and an adventurer, and he felt a sense of contentment and joy that he had never known before.
-            <br></br>
-            <br></br>
-            And so, as the adventurers sat around a campfire, drinking the beer that Bruinbrewinbeer had brewed, they knew that they had accomplished something great. They had overcome impossible odds, defeated powerful foes, and emerged stronger and more unified than ever before. They knew that whatever challenges lay ahead, they would face them together, with courage and determination, just like they had on this unforgettable journey.
-          </p>
-          <span>Tags</span> | <span>#brew</span>, <span>#dragons</span>, <span>#fiction</span>
-        </section>
         )}
```
Comment (nathanjslim) [inline client/src/pages/Blog.jsx]: Syntax error

**Comment by StephenODea54** [comment]
@nathanjslim I had a merge conflict and I tried to resolve it in the browser so that's where the two arrow functions stuff is probably coming from.

As for your other question, I have no idea lol.

**Comment by nathanjslim** [comment]
Okay its no worries; backend takes priority over a simple search function. 

I think I can re-rig the search function to search through your list of dictionaries as opposed to each div (which is what its currently doing).

**Comment by StephenODea54** [comment]
@nathanjslim which branch has your search functionality in it?

**Comment by nathanjslim** [comment]
Main currently has my search functionality in it.
No worries though, I can add it in later off your blog logic

**Comment by nathanjslim** [comment]
I removed it in the most recent commit to your branch and solved the syntax errors. Site works now. 
Will add search by EOD or tmrw.

**Comment by StephenODea54** [comment]
@nathanjslim oh oops my bad lol. Are you sure? I can abandon this PR and resubmit without messing your stuff up.

**Comment by nathanjslim** [comment]
@StephenODea54 nahh its fine lol, dw about it
shouldn't take too long anyways!

**Review by jpatterson933 — APPROVED** [review]

**Review by nathanjslim — APPROVED** [review]

**Pull request merged**